### PR TITLE
Interactive porep

### LINKED ISSUE #297: Interactive PoRep
This is a change that will likely be landing in the spec soon.

### PR DESCRIPTION
**Pull request by whyrusleeping**
Interactive porep by my best understanding of how it should work.

blocked on updating to the proofs code that supports this, and still needs some changes to plumb it all together (seal scheduler stuff)

should close #297

### CODE REVIEW

**Comment by whyrusleeping** [comment]
Should probably be combined with #432

File: build/params.go [hunk 1]
```diff
@@ -37,7 +37,7 @@ const PaymentChannelClosingDelay = 6 * 60 * 2 // six hours
 // Consensus / Network
 
 // Seconds
-const BlockDelay = 10
+const BlockDelay = 2
```
Comment (magik6k) [inline build/params.go]: Undo before merge

**Comment by magik6k** [comment]
TODO:
- [ ] Get deals to work
- [ ] Bubble up https://github.com/filecoin-project/rust-fil-proofs/pull/934
- [ ] Implement basic sectorbuilder task scheduler

File: storage/sector/store.go [hunk 2]
```diff
@@ -59,7 +60,32 @@ func (s *Store) SectorStatus(sid uint64) (*sectorbuilder.SectorSealingStatus, er
 	return &status, nil
 }
 
+func computePaddedSize(size uint64) uint64 {
+	// TODO: there is a better way to compute the log2 of an integer... i'm just lazy
+	logv := uint64(math.Log2(float64(size)))
```
Comment (magik6k) [inline storage/sector/store.go]: `64 - bits.LeadingZeros64(size)`

File: chain/actors/actor_miner.go [hunk 3]
```diff
@@ -199,44 +212,36 @@ func (sma StorageMinerActor) StorageMinerConstructor(act *types.Actor, vmctx typ
 	return nil, nil
 }
 
-type OnChainSealVerifyInfo struct {
+type SectorPreCommitInfo struct {
 	CommD []byte // TODO: update proofs code
```
Comment (magik6k) [inline chain/actors/actor_miner.go]: Check what to do here, probably remove

File: chain/actors/actor_storagemarket.go [hunk 4]
```diff
@@ -45,6 +47,7 @@ func (sma StorageMarketActor) Exports() []interface{} {
 		// 9: sma.SlashStorageDealCollateral,
 		// 10: sma.GetLastExpirationFromDealIDs,
 		11: sma.ActivateStorageDeals, // TODO: move under PublishStorageDeals after specs team approves
```
Comment (magik6k) [inline chain/actors/actor_storagemarket.go]: Check what updated spec does here

File: chain/actors/actor_storagemarket.go [hunk 5]
```diff
@@ -585,6 +588,51 @@ func transferFunds(from, to, amt types.BigInt) (types.BigInt, types.BigInt) {
...
+func (sma StorageMarketActor) ComputeDataCommitment(act *types.Actor, vmctx types.VMContext, params *ComputeDataCommitmentParams) ([]byte, ActorError) {
+	var self StorageMarketState
+	old := vmctx.Storage().GetHead()
+	if err := vmctx.Storage().Get(old, &self); err != nil {
+		return nil, err
+	}
+
+	deals, err := amt.LoadAMT(types.WrapStorage(vmctx.Storage()), self.Deals)
+	if err != nil {
+		// TODO: kind of annoying that this can be caused by gas, otherwise could be fatal
```
Comment (magik6k) [inline chain/actors/actor_storagemarket.go]: That's how it works, remove the comment

File: chain/deals/client_states.go [hunk 6]
```diff
@@ -126,16 +133,72 @@ func (c *Client) staged(ctx context.Context, deal ClientDeal) error {
 }
 
 func (c *Client) sealing(ctx context.Context, deal ClientDeal) error {
-	resp, err := c.readStorageDealResp(deal)
-	if err != nil {
-		return err
+
+	// TODO: disconnect
```
Comment (magik6k) [inline chain/deals/client_states.go]: Want

File: chain/deals/client_states.go [hunk 7]
```diff
@@ -126,16 +133,72 @@ func (c *Client) staged(ctx context.Context, deal ClientDeal) error {
...
+			}
+
+			return true, false, nil
+		}
+
+		return false, true, nil
+	}
+
+	called := func(msg *types.Message, ts *types.TipSet, curH uint64) (bool, error) {
+		// TODO: handle errors
```
Comment (magik6k) [inline chain/deals/client_states.go]: Want before merge

File: chain/deals/client_states.go [hunk 8]
```diff
@@ -126,16 +133,72 @@ func (c *Client) staged(ctx context.Context, deal ClientDeal) error {
...
+
+		if msg == nil {
+			log.Error("timed out waiting for deal activation... what now?")
+			return false, nil
+		}
+
+		// TODO: can check msg.Params to see if we should even bother checking the state
```
Comment (magik6k) [inline chain/deals/client_states.go]: Want before merge

File: api/api_full.go [hunk 9]
```diff
@@ -103,6 +74,7 @@ type FullNode interface {
 	// ClientImport imports file under the specified path into filestore
 	ClientImport(ctx context.Context, path string) (cid.Cid, error)
 	ClientStartDeal(ctx context.Context, data cid.Cid, miner address.Address, epochPrice types.BigInt, blocksDuration uint64) (*cid.Cid, error)
+	ClientGetDealInfo(context.Context, cid.Cid) (*DealInfo, error)
 	ClientListDeals(ctx context.Context) ([]DealInfo, error)
 	ClientHasLocal(ctx context.Context, root cid.Cid) (bool, error)
 	ClientFindData(ctx context.Context, root cid.Cid) ([]QueryOffer, error) // TODO: specify serialization mode we want (defaults to unixfs for now)
```
Comment (magik6k) [inline api/api_full.go]: Can remove this todo

File: chain/deals/client_states.go [hunk 10]
```diff
@@ -126,16 +133,72 @@ func (c *Client) staged(ctx context.Context, deal ClientDeal) error {
...
+		}
+
+		return false, nil
 	}
 
-	if resp.State != api.DealComplete {
-		return xerrors.Errorf("deal wasn't complete (State=%d)", resp.State)
+	revert := func(ctx context.Context, ts *types.TipSet) error {
+		log.Warn("deal activation reverted; TODO: actually handle this!")
+		// TODO: Just go back to DealSealing?
```
Comment (magik6k) [inline chain/deals/client_states.go]: Probably do that (check how that affects event handling when we return `false` for `more`)

File: chain/deals/provider_states.go [hunk 11]
```diff
@@ -221,89 +219,65 @@ func (p *Provider) staged(ctx context.Context, deal MinerDeal) (func(*MinerDeal)
...
-	}
-
-	return status, nil
-}
-
 func (p *Provider) sealing(ctx context.Context, deal MinerDeal) (func(*MinerDeal), error) {
 	err := p.sendSignedResponse(&Response{
 		State:    api.DealSealing,
 		Proposal: deal.ProposalCid,
+		// TODO: Send sector ID
```
Comment (magik6k) [inline chain/deals/provider_states.go]: Or deal id

File: chain/deals/provider_states.go [hunk 12]
```diff
@@ -221,89 +219,65 @@ func (p *Provider) staged(ctx context.Context, deal MinerDeal) (func(*MinerDeal)
...
 	if err != nil {
 		log.Warnf("Sending deal response failed: %s", err)
 	}
 
-	if err := p.secst.SealSector(ctx, deal.SectorID); err != nil {
+	log.Info("About to seal sector!", deal.ProposalCid, deal.SectorID)
+	if err := p.sminer.SealSector(ctx, deal.SectorID); err != nil {
 		return nil, xerrors.Errorf("sealing sector failed: %w", err)
 	}
+	// TODO: Let's not care after this point, for now at least, client can watch the chain
```
Comment (magik6k) [inline chain/deals/provider_states.go]: Disconnect here

File: chain/deals/provider_utils.go [hunk 13]
```diff
@@ -45,21 +44,21 @@ func (p *Provider) failDeal(id cid.Cid, cerr error) {
...
 		log.Errorw("failed to read proposal message", "error", err)
 		return proposal, err
 	}
 
-	if err := proposal.Verify(); err != nil {
+	if err := proposal.DealProposal.Verify(); err != nil {
 		return proposal, xerrors.Errorf("verifying StorageDealProposal: %w", err)
 	}
 
 	// TODO: Validate proposal maybe
```
Comment (magik6k) [inline chain/deals/provider_utils.go]: See what we are doing now, this might already be the case

File: chain/deals/client_utils.go [hunk 14]
```diff
@@ -66,7 +80,7 @@ func (c *Client) readStorageDealResp(deal ClientDeal) (*Response, error) {
 	// TODO: verify signature
```
Comment (magik6k) [inline chain/deals/client_utils.go]: Make sure we do that

File: chain/actors/actor_storagemarket.go (hunk 4 above)
Comment (magik6k) [reply chain/actors/actor_storagemarket.go]: Not much at all, let's keep this this way for now

File: storage/garbage.go [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/actors"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/lib/sectorbuilder"
+)
+
+// TODO: expected sector ID
```
Comment (magik6k) [inline storage/garbage.go]: Really need this

File: chain/actors/actor_miner.go [hunk 16]
```diff
@@ -257,10 +260,93 @@ func (sma StorageMinerActor) CommitSector(act *types.Actor, vmctx types.VMContex
...
+	if err != nil {
+		return nil, err
+	}
+
+	us, ok := self.PreCommittedSectors[uintToStringKey(params.SectorID)]
+	if !ok {
+		return nil, aerrors.New(1, "no pre-commitment found for sector")
+	}
+
+	if us.ReceivedEpoch+build.InteractivePoRepDelay > vmctx.BlockHeight() {
```
Comment (whyrusleeping) [inline chain/actors/actor_miner.go]: this should probably be >=, not sure how you can get randomness from the block that youre getting the message included in, and having time to compute the PoSt and all.

File: chain/deals/client_states.go [hunk 17]
```diff
@@ -1,142 +1,224 @@
...
-	pubmsg, err := c.chain.GetMessage(*resp.PublishMessage)
+	return func(info *ClientDeal) {
+		info.PublishMessage = resp.PublishMessage
+	}, nil
+}
+
+func (c *Client) accepted(ctx context.Context, deal ClientDeal) (func(*ClientDeal), error) {
+	log.Infow("DEAL ACCEPTED!")
+
+	pubmsg, err := c.chain.GetMessage(*deal.PublishMessage)
```
Comment (whyrusleeping) [inline chain/deals/client_states.go]: We might see this one failing with some frequency, the message isnt likely to propagate faster than the miner can send us an update directly

File: chain/deals/client_states.go [hunk 18]
```diff
@@ -1,142 +1,224 @@
...
+	}
+
+	called := func(msg *types.Message, ts *types.TipSet, curH uint64) (more bool, err error) {
+		defer func() {
+			if err != nil {
+				select {
+				case c.updated <- clientDealUpdate{
+					newState: api.DealComplete,
+					id:       deal.ProposalCid,
+					err:      xerrors.Errorf("handling applied event: %w", err),
```
Comment (whyrusleeping) [inline chain/deals/client_states.go]: why have an error here?

Comment (whyrusleeping) [reply chain/deals/client_states.go]: oh. i see. we're in the `if err != nil {`. duh.sorry.

File: chain/deals/client_states.go [hunk 19]
```diff
@@ -1,142 +1,224 @@
...
 
-	ok, err := sectorbuilder.VerifyPieceInclusionProof(ssize, deal.Proposal.Size, deal.Proposal.CommP, resp.CommD, resp.PieceInclusionProof.ProofElements)
-	if err != nil {
-		return xerrors.Errorf("verifying piece inclusion proof in staged deal %s: %w", deal.ProposalCid, err)
-	}
-	if !ok {
-		return xerrors.Errorf("verifying piece inclusion proof in staged deal %s failed", deal.ProposalCid)
-	}*/
+		if sd.ActivationEpoch == 0 {
+			return true, nil
```
Comment (whyrusleeping) [inline chain/deals/client_states.go]: how would this happen? If we found the deal in the message, but it ends up not getting activated... maybe this is an error?

File: chain/deals/client_states.go [hunk 20]
```diff
@@ -1,142 +1,224 @@
...
-func (c *Client) staged(ctx context.Context, deal ClientDeal) error {
-	/* miner seals our data, hopefully */
+		return false, true, nil
+	}
+
+	called := func(msg *types.Message, ts *types.TipSet, curH uint64) (more bool, err error) {
+		defer func() {
+			if err != nil {
+				select {
+				case c.updated <- clientDealUpdate{
```
Comment (whyrusleeping) [inline chain/deals/client_states.go]: this feels a little bit deadlocky to me, probably fine if theres a sufficient buffer, but still nervous

Comment (whyrusleeping) [reply chain/deals/client_states.go]: Oh, I guess these calls don't actually happen inside the client state machine loop, they happen over in the events system loop. Should be fine?

Comment (magik6k) [reply chain/deals/client_states.go]: That's a separate goroutine - https://github.com/filecoin-project/lotus/pull/472/files#diff-e264760819accf7406fce3d1ec61e7e1L16

**Comment by magik6k** [comment]
Last todos:

- [ ] Storage market fund manager
- [ ] API to expose sectorbuilder worker usage
  - + show in pond / miner info cmd
- [ ] Fix `SectorsList` / `SectorsStatus`
- [ ] Fix Circle

File: chain/deals/provider_states.go [hunk 21]
```diff
@@ -221,89 +180,31 @@ func (p *Provider) staged(ctx context.Context, deal MinerDeal) (func(*MinerDeal)
...
-	}
-
-	_, err = p.waitSealed(ctx, deal)
-	if err != nil {
-		return nil, err
-	}
-	// TODO: Spec doesn't say anything about inclusion proofs anywhere
-	//  Not sure what mechanisms prevents miner from storing data that isn't
-	//  clients' data
+	// TODO: consider waiting for seal to happen
```
Comment (whyrusleeping) [inline chain/deals/provider_states.go]: Are we fine leaving this unhandled for now?

File: chain/events/events_called.go [hunk 22]
```diff
@@ -146,21 +147,25 @@ func (e *calledEvents) checkNewCalls(ts *types.TipSet) {
 
 func (e *calledEvents) queueForConfidence(triggerId uint64, msg *types.Message, ts *types.TipSet) {
 	trigger := e.triggers[triggerId]
-	triggerH := ts.Height() + uint64(trigger.confidence)
+
+	// messages are not applied in the tipset they are included in
```
Comment (whyrusleeping) [inline chain/events/events_called.go]: We *could* use the `MessagesForTipSet` on the parent tipset to source messages for this, which would avoid us having to deal with the off-by-one bits here

Comment (whyrusleeping) [reply chain/events/events_called.go]: General comment, doesnt need to be addressed in this PR, but we need to audit how this code handles incoming messages, I want to make sure its using the deduplicated messages from the parent tipset, and not the individual block messages as they come in

File: chain/deals/provider_states.go (hunk 21 above)
Comment (magik6k) [reply chain/deals/provider_states.go]: Yes, that would be only useful for display purposes right now (will be more useful when we implement graceful restarting later)

File: chain/deals/client_states.go (hunk 17 above)
Comment (magik6k) [reply chain/deals/client_states.go]: We can put the message in the deal response the miner sends us. Does that make sense?

File: lib/sectorbuilder/files.go [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+func (sb *SectorBuilder) stagedSectorFile(sectorID uint64) (*os.File, error) {
+	return os.OpenFile(sb.stagedSectorPath(sectorID), os.O_RDWR|os.O_CREATE, 0644)
+}
+
+func (sb *SectorBuilder) sealedSectorPath(sectorID uint64) (string, error) {
+	path := filepath.Join(sb.sealedDir, sb.sectorName(sectorID))
+
+	e, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0644)
```
Comment (whyrusleeping) [inline lib/sectorbuilder/files.go]: why do we want to create the file here?

File: lib/sectorbuilder/sectorbuilder.go [hunk 24]
```diff
@@ -26,36 +37,120 @@ type SectorInfo = sectorbuilder.SectorInfo
...
 }
 
-type SectorBuilderConfig struct {
-	SectorSize  uint64
-	Miner       address.Address
+type Config struct {
+	SectorSize uint64
+	Miner      address.Address
+
+	WorkerThreads uint8
```
Comment (whyrusleeping) [inline lib/sectorbuilder/sectorbuilder.go]: only a uint8? but what about my 123512126 core machine?

File: peermgr/peermgr.go [hunk 25]
```diff
@@ -43,6 +44,7 @@ type PeerMgr struct {
 }
 
 func NewPeerMgr(h host.Host, dht *dht.IpfsDHT, bootstrap dtypes.BootstrapPeers) *PeerMgr {
+	bootstrap = nil
```
Comment (magik6k) [inline peermgr/peermgr.go]: Looks like someone was debugging

File: chain/deals/client_states.go (hunk 10 above)
Comment (magik6k) [reply chain/deals/client_states.go]: (Can fix later)

File: chain/events/events_called.go (hunk 22 above)
Comment (magik6k) [reply chain/events/events_called.go]: It handles the deduplication. Added a test just in case.

File: lib/sectorbuilder/sectorbuilder.go [hunk 26]
```diff
@@ -26,36 +37,120 @@ type SectorInfo = sectorbuilder.SectorInfo
...
+
+	CacheDir    string
 	SealedDir   string
 	StagedDir   string
 	MetadataDir string
 }
 
-func New(cfg *SectorBuilderConfig) (*SectorBuilder, error) {
+func New(cfg *Config, ds dtypes.MetadataDS, lc fx.Lifecycle) (*SectorBuilder, error) {
```
Comment (whyrusleeping) [inline lib/sectorbuilder/sectorbuilder.go]: It feels weird to depend on something from fx here, makes it more difficult to construct one of these in a test (even if it is okay to pass nil)

File: lib/sectorbuilder/sectorbuilder.go (hunk 24 above)
Comment (magik6k) [reply lib/sectorbuilder/sectorbuilder.go]: go-sectorbuilder limitation

File: lib/sectorbuilder/sectorbuilder.go [hunk 27]
```diff
@@ -68,31 +163,164 @@ func (sb *SectorBuilder) Destroy() {
...
+	stagedFile, err := sb.stagedSectorFile(sectorId)
+	if err != nil {
+		return PublicPieceInfo{}, err
+	}
+
+	_, _, commP, err := sectorbuilder.StandaloneWriteWithAlignment(f, pieceSize, stagedFile, existingPieceSizes)
+	if err != nil {
+		return PublicPieceInfo{}, err
+	}
+	/*if writeUnpadded != pieceSize {
```
Comment (whyrusleeping) [inline lib/sectorbuilder/sectorbuilder.go]: we should uncomment or delete this check

File: lib/sectorbuilder/sectorbuilder.go [hunk 28]
```diff
@@ -68,31 +163,164 @@ func (sb *SectorBuilder) Destroy() {
...
+		return RawSealPreCommitOutput{}, err
+	}
+
+	sealedPath, err := sb.sealedSectorPath(sectorID)
+	if err != nil {
+		return RawSealPreCommitOutput{}, err
+	}
+
+	var sum uint64
+	for _, piece := range pieces {
```
Comment (whyrusleeping) [inline lib/sectorbuilder/sectorbuilder.go]: these pieces include the padded size, correct?

Comment (magik6k) [reply lib/sectorbuilder/sectorbuilder.go]: Yes, that's the `UserBytes(NearestPowerOf2(size))` size

File: chain/deals/client_states.go (hunk 17 above)
Comment (whyrusleeping) [reply chain/deals/client_states.go]: Yeah, that would be good I think

Comment (whyrusleeping) [reply chain/deals/client_states.go]: We could also just do the wait first, before this validation.

File: storage/garbage.go [hunk 29]
```diff
@@ -0,0 +1,138 @@
...
+	}
+
+	params, aerr := actors.SerializeParams(&actors.PublishStorageDealsParams{
+		Deals: deals,
+	})
+	if aerr != nil {
+		return nil, xerrors.Errorf("serializing PublishStorageDeals params failed: ", aerr)
+	}
+
+	// TODO: We may want this to happen after fetching data
```
Comment (whyrusleeping) [inline storage/garbage.go]: fetching what data?

File: storage/garbage.go [hunk 30]
```diff
@@ -0,0 +1,138 @@
...
+			Size:   ppi.Size,
+			CommP:  ppi.CommP[:],
+		}
+	}
+
+	return out, nil
+}
+
+func (m *Miner) StoreGarbageData(_ context.Context) error {
+	ctx := context.TODO()
```
Comment (whyrusleeping) [inline storage/garbage.go]: context.TODO, and ignoring the context in the function parameter?

Comment (magik6k) [reply storage/garbage.go]: The function ctx gets cancelled when the cmd returns

File: storage/post.go [hunk 31]
```diff
@@ -96,87 +99,184 @@ func (m *Miner) scheduleNextPost(ppe uint64) {
...
+	ctx, span := trace.StartSpan(ctx, "storage.preparePost")
+	defer span.End()
+
+	sset, err := p.m.api.StateMinerProvingSet(ctx, p.m.maddr, p.ts)
+	if err != nil {
+		return xerrors.Errorf("failed to get proving set for miner: %w", err)
+	}
+	p.sset = sset
+
+	r, err := p.m.api.ChainGetRandomness(ctx, p.ts, nil, int(int64(p.ts.Height())-int64(p.ppe)+int64(build.PoStChallangeTime)+int64(build.PoStRandomnessLookback))) // TODO: review: check math
```
Comment (whyrusleeping) [inline storage/post.go]:
```suggestion
        // Compute how many blocks back we have to look from the given tipset for the PoSt challenge
        challengeLookback := int((int64(p.ts.Height())-int64(p.ppe))+int64(build.PoStChallangeTime)+int64(build.PoStRandomnessLookback))
	r, err := p.m.api.ChainGetRandomness(ctx, p.ts, nil, challengeLookback)
```

File: storage/sealing.go [hunk 32]
```diff
@@ -0,0 +1,253 @@
...
+		m.handle(ctx, sector, m.finishPacking, api.Unsealed)
+	case api.Unsealed:
+		m.handle(ctx, sector, m.sealPreCommit, api.PreCommitting)
+	case api.PreCommitting:
+		m.handle(ctx, sector, m.preCommit, api.PreCommitted)
+	case api.PreCommitted:
+		m.handle(ctx, sector, m.preCommitted, api.SectorNoUpdate)
+	case api.Committing:
+		m.handle(ctx, sector, m.committing, api.Proving)
+	}
```
Comment (whyrusleeping) [inline storage/sealing.go]: I'd feel more comfortable if there was a default here that logged that something weird was going on

File: storage/sealing.go [hunk 33]
```diff
@@ -0,0 +1,253 @@
...
+}
+
+func (m *Miner) failSector(id uint64, err error) {
+	log.Errorf("sector %d error: %+v", id, err)
+	panic(err) // todo: better error handling strategy
```
Comment (whyrusleeping) [inline storage/sealing.go]: probably want to do something better than panic, for now, maybe 'not panic' is acceptable?

File: storage/sealing.go [hunk 34]
```diff
@@ -0,0 +1,253 @@
...
+	if err != nil {
+		return 0, xerrors.Errorf("adding piece to sector: %w", err)
+	}
+
+	return sid, m.newSector(ctx, sid, dealID, ref, ppi)
+}
+
+func (m *Miner) newSector(ctx context.Context, sid uint64, dealID uint64, ref string, ppi sectorbuilder.PublicPieceInfo) error {
+	si := &SectorInfo{
+		State:    api.UndefinedSectorState,
```
Comment (whyrusleeping) [inline storage/sealing.go]: undefined sector state? whats the purpose of doing this?

File: storage/sealing_utils.go [hunk 35]
```diff
@@ -0,0 +1,34 @@
+package storage
+
+import (
+	"math/bits"
+
+	"github.com/filecoin-project/lotus/lib/sectorbuilder"
+)
+
+func fillersFromRem(toFill uint64) ([]uint64, error) {
+	toFill += toFill / 127 // convert to in-sector bytes for easier math
```
Comment (whyrusleeping) [inline storage/sealing_utils.go]: is that... accurate?

Comment (whyrusleeping) [reply storage/sealing_utils.go]: cc @dignifiedquire

File: storage/sealing_utils.go [hunk 36]
```diff
@@ -0,0 +1,34 @@
...
+
+	out := make([]uint64, bits.OnesCount64(toFill))
+	for i := range out {
+		next := bits.TrailingZeros64(toFill)
+		psize := uint64(1) << next
+		toFill ^= psize
```
Comment (whyrusleeping) [inline storage/sealing_utils.go]: what even is this math @magik6k

File: storage/sector_states.go [hunk 37]
```diff
@@ -0,0 +1,231 @@
...
+		return nil, err
+	}
+
+	if mw.Receipt.ExitCode != 0 {
+		log.Error("sector precommit failed: ", mw.Receipt.ExitCode)
+		return nil, err
+	}
+	log.Info("precommit message landed on chain: ", sector.SectorID)
+
+	randHeight := mw.TipSet.Height() + build.InteractivePoRepDelay - 1 // -1 because of how the messages are applied
```
Comment (whyrusleeping) [inline storage/sector_states.go]: the -1 seems like an odd constraint. We should probably figure that out and remove it to keep the math clean

File: storage/garbage.go (hunk 30 above)
Comment (whyrusleeping) [reply storage/garbage.go]: oh, thats kinda annoying...

Comment (whyrusleeping) [reply storage/garbage.go]: probably should add a comment letting future us know that

File: lib/sectorbuilder/files.go (hunk 23 above)
Comment (magik6k) [reply lib/sectorbuilder/files.go]: Sectorbuilder doesn't do that for us (78% sure here)

File: storage/sealing.go (hunk 33 above)
Comment (magik6k) [reply storage/sealing.go]: Yes, that's marginally better

File: storage/sealing.go (hunk 34 above)
Comment (magik6k) [reply storage/sealing.go]: `onIncoming` sets this to `Packing`after the sector arrives in state manager

File: storage/sector_states.go (hunk 37 above)
Comment (magik6k) [reply storage/sector_states.go]: The correct thing would be to make the `vmctx.BlockHeight()` be the height at which the message is computed instead of the height at which it it included. That's probably tricky, and a bit out of scope here.
